### PR TITLE
Issue: Mega Nav Caching

### PR DESCRIPTION
**Pull request by mufudzimasaire**
## Description

Implements caching for menus. This implementation caches the menus for a set duration - configured via the environment variable `MENUS_CACHE_DURATION`. Calls for menus that are made within/during this duration are served the cached menus, thus no external API calls are made.

After this duration elapses, an API call is made to the platform and the response is stored in the cache store (default: memcached) and the cycle continues.

Note: The Heroku  & RedisLabs recommended memcache client library `memjs` has a DoS security vulnerability.  Ticket has been added in their repo and ours: https://github.com/shiftcommerce/trendy-golf/issues/265

## Related Issue

https://github.com/shiftcommerce/trendy-golf/issues/131 

### Checklist

- [x] Appropriate commenting on functions/components (JSDoc format).
- [x] All SHIFT library versions bumped and published (utilised in this PR).
- [x] Unit tests added/amended.
- [x] Integration tests added/amended.
- [x] Environment variables supplied (.env.example also updated)

### How has/can this been tested?

- [x] Serverside rendering
- [x] Clientside rendering

Manually tested on heroku by setting a low cache ttl - 

qa on heroku

### Dependencies Review

The in-memory store - Memcached is introduced in this PR. The environment variables added include:

```bash
MEMCACHE_SERVERS=<'127.0.0.1:11211,192.168.0.102:11211,192.168.0.103:11211'>
MEMCACHE_USERNAME=<memcache_server_username>
MEMCACHE_PASSWORD=<memcache_server_password>
MENUS_CACHE_DURATION=<the_duration_the_menus_should_be_cached_in_seconds>
```

If provisioning memcache service on heroku eg. (memcached cloud or memcachier), the above server and user name envs do not need to be set as the following env vars will be read automatically:

```
MEMCACHEDCLOUD_SERVERS
MEMCACHEDCLOUD_USERNAME
MEMCACHEDCLOUD_PASSWORD

OR

MEMCACHIER_SERVERS
MEMCACHIER_USERNAME
MEMCACHIER_PASSWORD
```

### CODE REVIEW

File: packages/shift-next-routes/src/lib/menu-cache.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      return {}
+    }
+  }
+
+  /**
+   * Sets the received API menu response
+   * @param {object} response - menu API response
+   * @param {number} cacheDuration - the cache duration in seconds
+   */
+  async set (response, cacheDuration = process.env.MENUS_CACHE_DURATION) {
```
Comment (cbarratt) [inline packages/shift-next-routes/src/lib/menu-cache.js]: Can we set a default for this to guard against this not being set? :)

File: packages/shift-next-routes/src/lib/menu-cache.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+   * Sets the received API menu response
+   * @param {object} response - menu API response
+   * @param {number} cacheDuration - the cache duration in seconds
+   */
+  async set (response, cacheDuration = (process.env.MENUS_CACHE_DURATION || 300)) {
+    try {
+      const result = await this.cache.set(this.cacheKey, JSON.stringify(response), { expires: cacheDuration })
+      return result
+    } catch (error) {
+      console.log('Error setting menu cache', error)
```
Comment (marcgear) [inline packages/shift-next-routes/src/lib/menu-cache.js]: can we get this to be a logger.error() line somehow?

File: packages/shift-next-routes/src/express/menu-handler.js [hunk 3]
```diff
@@ -18,3 +32,12 @@ module.exports = {
     }
   }
 }
+
+const fetchMenuFromAPI = async (req, res) => {
+  // fetch menus from API
+  let response = await SHIFTClient.getMenusV1(req.query)
```
Comment (cbarratt) [inline packages/shift-next-routes/src/express/menu-handler.js]:
```suggestion
  const response = await SHIFTClient.getMenusV1(req.query)
```

File: packages/shift-next-routes/src/express/menu-handler.js [hunk 4]
```diff
@@ -1,11 +1,25 @@
 const { SHIFTClient } = require('@shiftcommerce/shift-node-api')
+const MenuCache = require('../lib/menu-cache')
 const { setSurrogateHeaders } = require('../lib/set-cache-headers')
 
 module.exports = {
-  getMenu: async (req, res) => {
-    const response = await SHIFTClient.getMenusV1(req.query)
-
-    setSurrogateHeaders(response.headers, res)
```
Comment (marcgear) [inline packages/shift-next-routes/src/express/menu-handler.js]: we still need to set the cache keys for the menu - doing this only for times we go to the API will prevent them from making it in

File: packages/shift-next-routes/src/lib/menu-cache.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  /**
+   * Reads the cached menu response
+   * @return {object} - the cached menu API response
+   */
+  async read () {
+    try {
+      const data = await this.cache.get(this.cacheKey)
+      return data ? JSON.parse(data.value) : {}
+    } catch (error) {
+      console.error('Error fetching menu cache', error)
```
Comment (marcgear) [inline packages/shift-next-routes/src/lib/menu-cache.js]: same with this log line

File: packages/shift-next-routes/src/lib/menu-cache.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      return {}
+    }
+  }
+
+  /**
+   * Sets the received API menu response
+   * @param {object} response - menu API response
+   * @param {number} cacheDuration - the cache duration in seconds
+   */
+  async set (response, cacheDuration = (process.env.MENUS_CACHE_DURATION || 300)) {
```
Comment (cbarratt) [inline packages/shift-next-routes/src/lib/menu-cache.js]: process.env.MENUS_CACHE_DURATION - do we need to parse this as int when it comes form ENV?

**Pull request merged**